cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/MultipleOutputs.java

```diff
@@ -242,6 +245,47 @@ private static Job getNamedJob(JobContext context, String namedOutput) throws IO
     return job;
   }
 
+  /**
+   * Wraps RecordWriter to increment output counters.
+   */
+  @SuppressWarnings("unchecked")
+  private static class RecordWriterWithCounter extends RecordWriter {
+    private final RecordWriter writer;
+    private final String groupName;
+    private final String counterName;
+    private final TaskInputOutputContext context;
+
+    public RecordWriterWithCounter(RecordWriter writer, TaskInputOutputContext context) {
+      this.writer = writer;
+      this.context = context;
+      this.groupName = TaskCounter.class.getName();
+      this.counterName = getCounterName(context);
+    }
+
+    @SuppressWarnings("unchecked")
+    public void write(Object key, Object value) throws IOException, InterruptedException {
+      context.getCounter(groupName, counterName).increment(1);
+      writer.write(key, value);
+    }
+
+    public void close(TaskAttemptContext context) throws IOException, InterruptedException {
+      writer.close(context);
+    }
+
+    private String getCounterName(TaskInputOutputContext context) {
+      MapReduceMetrics.TaskType taskType = MapReduceMetrics.TaskType.from(context.getTaskAttemptID().getTaskType());
+      Preconditions.checkNotNull(taskType, "Illegal task type: " + context.getTaskAttemptID().getTaskType());
```

Comment:
1. For the "Enum.from" method, the common convention is to throw IllegalArgumentException if the string type name is not recognizable. 
2. For the use case that you pasted above, it's more like the enum need to have another method (e.g. `boolean TaskType.hasType(String typeString)`).